### PR TITLE
Deserialize BlobStore Metadata Files in a Streaming Manner

### PR DESCRIPTION
**Pull request by original-brownbear**
We were reading the full file contents up-front here because of the complexity
of verifying the footer otherwise. This commit moves the logic for reading metadata
blobs (that can become quite sizable in some cases + there's plans for larger aggregate meta blobs as well) in a streaming manner by manually doing the footer verification as Lucene's utility methods don't allow for
verification on top of a stream.

A possible follow-up to this would be to fix the write side the same way and get rid of the need to fully-buffer blobs before writing there as well.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/snapshots/BlobStoreFormatTests.java [hunk 1]
```diff
@@ -127,16 +126,14 @@ public void testBlobCorruption() throws IOException {
...
-        assertEquals(checksumFormat.read(blobContainer, "test-path", xContentRegistry(), MockBigArrays.NON_RECYCLING_INSTANCE).getText(),
+        assertEquals(checksumFormat.read(blobContainer, "test-path", xContentRegistry()).getText(),
                 testString);
         randomCorruption(blobContainer, "test-path");
         try {
-            checksumFormat.read(blobContainer, "test-path", xContentRegistry(), MockBigArrays.NON_RECYCLING_INSTANCE);
+            checksumFormat.read(blobContainer, "test-path", xContentRegistry());
             fail("Should have failed due to corruption");
-        } catch (ElasticsearchCorruptionException ex) {
-            assertThat(ex.getMessage(), containsString("test-path"));
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/snapshots/BlobStoreFormatTests.java:137]: No need to include the path here and complicate the code IMO, we wrap and/or log the path of what failed to deserialize upstream anyway to get insight into where other `IOException`s happened.

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-2

File: server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java [hunk 2]
```diff
@@ -93,40 +92,166 @@ public ChecksumBlobStoreFormat(String codec, String blobNameFormat, CheckedFunct
...
+     * a parser backed by this stream will only see the blob's body.
+     */
+    private static final class DeserializeMetaBlobInputStream extends FilterInputStream {
+
+        // checksum updated with all but the last 8 bytes read from the wrapped stream
+        private final CRC32 crc32 = new CRC32();
+
+        private final byte[] buffer = new byte[1024 * 8];
+
+        private int bufferCount;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java:142]:
```suggestion
        // the number of bytes in the buffer, in [0, buffer.length], equal to buffer.length unless the last fill hit EOF
        private int bufferCount;
```

File: server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java [hunk 3]
```diff
@@ -93,40 +92,166 @@ public ChecksumBlobStoreFormat(String codec, String blobNameFormat, CheckedFunct
...
+
+        private int bufferPos;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java:144]:
```suggestion
        // the current read position within the buffer, in [0, bufferCount)
        private int bufferPos;
```

File: server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java [hunk 4]
```diff
@@ -93,40 +92,166 @@ public ChecksumBlobStoreFormat(String codec, String blobNameFormat, CheckedFunct
...
+     * Wrapper input stream for deserializing blobs that come with a Lucene header and footer in a streaming manner. It manually manages
+     * a read buffer to enable not reading into the last 16 bytes (the footer length) of the buffer via the standard read methods so that
+     * a parser backed by this stream will only see the blob's body.
+     */
+    private static final class DeserializeMetaBlobInputStream extends FilterInputStream {
+
+        // checksum updated with all but the last 8 bytes read from the wrapped stream
+        private final CRC32 crc32 = new CRC32();
+
+        private final byte[] buffer = new byte[1024 * 8];
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java:140]:
```suggestion
        // Only the first buffer.length - 16 bytes are exposed by the read() methods; once the read position reaches 16 bytes from the end
        // of the buffer the remaining 16 bytes are moved to the start of the buffer and the rest of the buffer is filled from the stream.
        private final byte[] buffer = new byte[1024 * 8];
```

File: server/src/test/java/org/elasticsearch/snapshots/BlobStoreFormatTests.java [hunk 5]
```diff
@@ -98,10 +97,10 @@ public void testBlobStoreOperations() throws IOException {
                 MockBigArrays.NON_RECYCLING_INSTANCE);
 
         // Assert that all checksum blobs can be read
-        assertEquals(checksumSMILE.read(blobContainer, "check-smile", xContentRegistry(), MockBigArrays.NON_RECYCLING_INSTANCE).getText(),
+        assertEquals(checksumSMILE.read(blobContainer, "check-smile", xContentRegistry()).getText(),
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/snapshots/BlobStoreFormatTests.java:100]: I think these tests only cover very small blobs so we're not really exercising the corners of the refilling logic. Let's have some blobs that are up to a few times larger than the buffer size too.

We also apparently only use a few different read sizes and only call the one-byte `read()` for reading the header.

IMO `DeserializeMetaBlobInputStream` could reasonably be a top-level class with some more focussed tests.

**Review by DaveCTurner — COMMENTED** [review]
Seems nicer indeed. I left some ideas mainly on comments, assertions and tests.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java [hunk 6]
```diff
@@ -93,40 +92,166 @@ public ChecksumBlobStoreFormat(String codec, String blobNameFormat, CheckedFunct
...
+            return bufferCount - bufferPos - CodecUtil.footerLength();
+        }
+
+        private void fill() throws IOException {
+            if (bufferCount == 0) {
+                bufferCount = Streams.readFully(in, buffer, 0, buffer.length);
+            } else {
+                // crc and discard all but the last 16 bytes in the buffer that might be the footer bytes
+                final int footerLen = CodecUtil.footerLength();
+                assert bufferCount >= footerLen;
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java:250]: ++ adjusted as suggested now

File: server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java:145]: ++ added that comment

File: server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java:148]: ++ added

File: server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java (hunk 4 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java:142]: ++ added

File: server/src/test/java/org/elasticsearch/snapshots/BlobStoreFormatTests.java (hunk 5 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/BlobStoreFormatTests.java]: > I think these tests only cover very small blobs so we're not really exercising the corners of the refilling logic. Let's have some blobs that are up to a few times larger than the buffer size too.

++ I adjusted the test to use some larger blobs now.

> We also apparently only use a few different read sizes and only call the one-byte read() for reading the header.

Right, it's just single reads during header read, 4k for the decompressing stream buffer and exactly 8000 from the Jackson SMILE parser at the moment as far as I can tell. Testing with random sizes of up to 3x8k should run into all mathematically possible corner cases now I think. At least I was able to run 100k+ iterations without failure with the new tests.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java [hunk 7]
```diff
@@ -93,40 +92,166 @@ public ChecksumBlobStoreFormat(String codec, String blobNameFormat, CheckedFunct
...
+            bufferPos += read;
+            return read;
+        }
+
+        /**
+         * Verify footer of the bytes read by this stream the same way {@link CodecUtil#checkFooter(ChecksumIndexInput)} would.
+         *
+         * @throws CorruptStateException if footer is found to be corrupted
+         */
+        void verifyFooter() throws CorruptStateException {
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java:208]: Hmm I think I like the current version better but maybe I'm missing something. It seems like we would only be able to simplify away some `bufferPos + ` with this right? We still have to separately call the `crc32.update` for the first 8 footer bytes so idk :), this version seems simpler but it's not a strong preference either.

**Comment by original-brownbear** [comment]
Thanks @DaveCTurner :) All suggestions applied with the exception of extracting the stream to a top-level class. It seemed for now just testing a wider range of blob sizes gives us the same coverage with less noise.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java:208]: FWIW I prefer the current way, which makes it more explicit that the first 8 bytes of the footer are also part of the checksum.

File: server/src/test/java/org/elasticsearch/snapshots/BlobStoreFormatTests.java [hunk 8]
```diff
@@ -93,15 +82,16 @@ public void testBlobStoreOperations() throws IOException {
         ChecksumBlobStoreFormat<BlobObj> checksumSMILE = new ChecksumBlobStoreFormat<>(BLOB_CODEC, "%s", BlobObj::fromXContent);
 
         // Write blobs in different formats
-        checksumSMILE.write(new BlobObj("checksum smile"), blobContainer, "check-smile", false, MockBigArrays.NON_RECYCLING_INSTANCE);
-        checksumSMILE.write(new BlobObj("checksum smile compressed"), blobContainer, "check-smile-comp", true,
+        final String randomText = randomAlphaOfLengthBetween(0, 1024 * 8 * 3);
+        final String normalText = "checksum smile: " + randomText;
+        checksumSMILE.write(new BlobObj(normalText), blobContainer, "check-smile", false, MockBigArrays.NON_RECYCLING_INSTANCE);
+        final String compresedText = "checksum smile compressed: " + randomText;
```
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/snapshots/BlobStoreFormatTests.java:88]:
```suggestion
        final String compressedText = "checksum smile compressed: " + randomText;
```

**Review by tlrx — APPROVED** [review]
LGTM

File: server/src/test/java/org/elasticsearch/snapshots/BlobStoreFormatTests.java (hunk 5 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/snapshots/BlobStoreFormatTests.java]: I'm still concerned about coverage, for instance we apparently never hit the interesting cases in `read()`:

```
diff --git a/server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java b/server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java
index 7d43b3d715e..02b8aba2fd0 100644
--- a/server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java
+++ b/server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java
@@ -154,9 +154,11 @@ public final class ChecksumBlobStoreFormat<T extends ToXContent> {
         @Override
         public int read() throws IOException {
             if (buffered() <= 0) {
+                assert bufferCount == 0;
                 fill();
             }
             if (buffered() <= 0) {
+                assert false;
                 return -1;
             }
             return buffer[bufferPos++];
```

Fine for now but the scope for future bugs worries me.

I think I'd be semi-happy if we consolidated the logic that tries to make sure some bytes are available, something like this (with a suitable renaming too):

```
diff --git a/server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java b/server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java
index 7d43b3d715e..8b3738ab141 100644
--- a/server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java
+++ b/server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java
@@ -153,10 +153,7 @@ public final class ChecksumBlobStoreFormat<T extends ToXContent> {

         @Override
         public int read() throws IOException {
-            if (buffered() <= 0) {
-                fill();
-            }
-            if (buffered() <= 0) {
+            if (getAvailable() <= 0) {
                 return -1;
             }
             return buffer[bufferPos++];
@@ -187,10 +184,7 @@ public final class ChecksumBlobStoreFormat<T extends ToXContent> {
         }

         private int doRead(byte[] b, int off, int len) throws IOException {
-            if (buffered() <= 0) {
-                fill();
-            }
-            final int available = buffered();
+            final int available = getAvailable();
             if (available < 0) {
                 return -1;
             }
@@ -237,24 +231,26 @@ public final class ChecksumBlobStoreFormat<T extends ToXContent> {
             return CompressorFactory.COMPRESSOR.isCompressed(new BytesArray(buffer, bufferPos, bufferCount - bufferPos));
         }

-        private int buffered() {
-            // bytes in the buffer minus 16 bytes that could be the footer
-            return bufferCount - bufferPos - CodecUtil.footerLength();
-        }
-
-        private void fill() throws IOException {
+        /**
+         * @return the number of bytes available in the buffer, possibly refilling the buffer if needed
+         */
+        private int getAvailable() throws IOException {
+            final int footerLen = CodecUtil.footerLength();
             if (bufferCount == 0) {
+                // first read, fill the buffer
+                assert bufferPos == 0;
                 bufferCount = Streams.readFully(in, buffer, 0, buffer.length);
-            } else {
+            } else if (bufferPos == bufferCount - footerLen) {
                 // crc and discard all but the last 16 bytes in the buffer that might be the footer bytes
-                final int footerLen = CodecUtil.footerLength();
                 assert bufferCount >= footerLen;
-                assert bufferPos == bufferCount - footerLen;
                 crc32.update(buffer, 0, bufferPos);
                 System.arraycopy(buffer, bufferPos, buffer, 0, footerLen);
                 bufferCount = footerLen + Streams.readFully(in, buffer, footerLen, buffer.length - footerLen);
                 bufferPos = 0;
             }
+
+            // bytes in the buffer minus 16 bytes that could be the footer
+            return bufferCount - bufferPos - footerLen;
         }
     }
```

File: server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java:208]: Yeah on reflection I agree too. I thought this was a valid assertion, then I discovered it wasn't and tried to fix it up but basically it's better like this.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I think we have a problem in `BlobStoreFormatTests#testBlobCorruption`, since we now start reading the blob before checking it's valid so we get all sorts of other exceptions too.

File: server/src/test/java/org/elasticsearch/snapshots/BlobStoreFormatTests.java (hunk 5 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/BlobStoreFormatTests.java]: ++ applied that change.

**Comment by original-brownbear** [comment]
> I think we have a problem in BlobStoreFormatTests#testBlobCorruption, since we now start reading the blob before checking it's valid so we get all sorts of other exceptions too.

Urgh, nice find :) I partly fixed this by expanding the try-catch scope to include the header check and partly by just expecting more exceptions in the test. I didn't want to blanket catch-rethrow the ones I only added in the test because there's not "proof" of corruption with those and they may be the result of random bugs in our parsing (e.g. when we moved to not allow duplicate fields any more and would start throwing all kinds ).

**Review by DaveCTurner — APPROVED** [review]
:+1: LGTM thanks for the extra iterations

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
Hmm actually no I've had a change of heart. I think throwing these random exceptions on corruption will cause pain since they're indistinguishable from an actual corruption. Can we make it so that we read to the end in case of exception and report if the checksum was broken in any case?

**Comment by original-brownbear** [comment]
@DaveCTurner fair point :)

I pushed https://github.com/elastic/elasticsearch/pull/73149/commits/9c7fb6e9a28aece24e70e3dfcaf4acefd2235111 to always verify footer

**Review by DaveCTurner — APPROVED** [review]
Thanks, yes that's better IMO. LGTM now.

**Comment by original-brownbear** [comment]
Thanks David & Tanguy!

**Pull request merged**